pkg/verification/service/verification_service.go

```diff
@@ -0,0 +1,265 @@
+	ts, parseErr := time.Parse(TimestampLayout, signup.Annotations[v1alpha1.UserSignupVerificationInitTimestampAnnotationKey])
+	if parseErr != nil || (parseErr == nil && now.After(ts.Add(24*time.Hour))) {
+		// Set a new timestamp
+		signup.Annotations[v1alpha1.UserSignupVerificationInitTimestampAnnotationKey] = now.Format(TimestampLayout)
+		signup.Annotations[v1alpha1.UserSignupVerificationCounterAnnotationKey] = "0"
+	}
+
+	// get the verification counter (i.e. the number of times the user has initiated phone verification within
+	// the last 24 hours)
+	verificationCounter := signup.Annotations[v1alpha1.UserSignupVerificationCounterAnnotationKey]
+	var counter int
+	dailyLimit := s.config.GetVerificationDailyLimit()
+	if verificationCounter == "" {
+		counter = 0
+	} else {
+		counter, err = strconv.Atoi(verificationCounter)
+		if err != nil {
+			// We shouldn't get an error here, but if we do, we should probably set verification counter to the daily
+			// limit so that we at least now have a valid value
+			signup.Annotations[v1alpha1.UserSignupVerificationCounterAnnotationKey] = strconv.Itoa(dailyLimit)
+			counter = dailyLimit
+		}
+	}
+
+	// check if counter has exceeded the limit of daily limit - if at limit error out
+	if counter >= dailyLimit {
+		log.Error(ctx, err, fmt.Sprintf("%d attempts made. the daily limit of %d has been exceeded", counter, dailyLimit))
+		initError = errors.NewForbiddenError("daily limit exceeded", "cannot generate new verification code")
+	}
+
+	if initError == nil {
+		// generate verification code
+		verificationCode := generateVerificationCode()
+
+		// set the usersignup annotations
+		signup.Annotations[v1alpha1.UserSignupVerificationCounterAnnotationKey] = strconv.Itoa(counter + 1)
+		signup.Annotations[v1alpha1.UserSignupVerificationCodeAnnotationKey] = verificationCode
+		signup.Annotations[v1alpha1.UserVerificationExpiryAnnotationKey] = now.Add(
+			time.Duration(s.config.GetVerificationCodeExpiresInMin()) * time.Minute).Format(TimestampLayout)
+
+		// Generate the verification message with the new verification code
+		content := fmt.Sprintf(s.config.GetVerificationMessageTemplate(), verificationCode)
+		client := twilio.NewClient(s.config.GetTwilioAccountSID(), s.config.GetTwilioAuthToken(), s.HttpClient)
+		msg, err := client.Messages.SendMessage(s.config.GetTwilioFromNumber(), e164PhoneNumber, content, nil)
+		if err != nil {
+			log.Error(ctx, err, fmt.Sprintf("error while sending, code: %d message: %s", msg.ErrorCode, msg.ErrorMessage))
+
+			// If we get an error here then just die, don't bother updating the UserSignup
+			return errors.NewInternalError(err, "error while sending verification code")
+		}
+	}
+
+	_, updateErr := s.Services().SignupService().UpdateUserSignup(signup)
+	if updateErr != nil {
+		log.Error(ctx, err, "error while updating UserSignup resource")
+		initError = errors.NewInternalError(updateErr, "error while updating UserSignup resource")
```

Comment:
Still don't like the named returns (IMO it hurts code readability) but we can keep it if you really prefer using them ;)

Reply:
I did some more reading up on named return parameters and I don't think there's anything to be afraid of when using them, in fact most people said they helped by documenting the meaning of the return values.  Anyway, I've changed this to use an error variable that's declared up front instead.

Reply:
This is why I didn't insist :) Not sure if my code readability point is justified. Maybe it's just a personal preference.